lib/auth/sessions.go

```diff
@@ -111,21 +111,39 @@ func (s *Server) CreateAppSession(ctx context.Context, req types.CreateAppSessio
 // WaitForAppSession will block until the requested application session shows up in the
 // cache or a timeout occurs.
 func WaitForAppSession(ctx context.Context, sessionID, user string, ap ReadProxyAccessPoint) error {
-	_, err := ap.GetAppSession(ctx, types.GetAppSessionRequest{SessionID: sessionID})
+	return waitForWebSession(ctx, sessionID, user, types.KindAppSession, ap.NewWatcher, func(ctx context.Context, sessionID string) (types.WebSession, error) {
+		return ap.GetAppSession(ctx, types.GetAppSessionRequest{SessionID: sessionID})
+	})
+}
+
+// WaitForSnowflakeSession waits until the requested Snowflake session shows up int the cache
+// or a timeout occurs.
+func WaitForSnowflakeSession(ctx context.Context, sessionID, user string, ap SnowflakeSessionWatcher) error {
+	return waitForWebSession(ctx, sessionID, user, types.KindSnowflakeSession, ap.NewWatcher, func(ctx context.Context, sessionID string) (types.WebSession, error) {
+		return ap.GetSnowflakeSession(ctx, types.GetAppSessionRequest{SessionID: sessionID})
+	})
+}
+
+// waitForWebSession is an implementation for web session wait functions.
+func waitForWebSession(ctx context.Context, sessionID, user string, evenSubKind string,
+	newWatcherFn func(ctx context.Context, watch types.Watch) (types.Watcher, error),
+	getSessionFn func(ctx context.Context, sessionID string) (types.WebSession, error),
+) error {
+	_, err := getSessionFn(ctx, sessionID)
```

Comment:
I like the idea, but when I tried to changed I see a few more issues:
1. in case of errors I do not have access to logger. I can create it in `WaitForXSession()`, but then I need to pass it down to `waitForWebSession()` or create a new logger there?
2. `GetSession()` is called in two places in `waitForWebSession()`. Replacing the second usage it not so simple (or I don't see it)

I had another idea to create interface like
```go
type SessionWatcher interface {
	// NewWatcher returns a new event watcher.
	NewWatcher(ctx context.Context, watch types.Watch) (types.Watcher, error)
	// GetSessionByID returns WebSession by ID.
	GetSessionByID(ctx context.Context, ID string) (types.WebSession, error)
}
```
But then `GetSessionByID()` implementation is another copy-paste/looks as my local callback here.

If i figure out something more elegant I'll use it, otherwise I'll merge it as it is.
I also think this would be a good candidate for generics 😅